spylib/token.py

```diff
@@ -0,0 +1,428 @@
+from abc import ABC, abstractmethod
+from asyncio import sleep
+from datetime import datetime, timedelta
+from time import monotonic
+from typing import Any, Dict, List, Optional
+import json
+from loguru import logger
+from pydantic import BaseModel, root_validator, MissingError
+from pydantic.dataclasses import dataclass
+from pydantic.main import validate_model
+from shortuuid.main import ShortUUID  # type: ignore
+
+from math import ceil, floor
+from httpx import AsyncClient, Response
+from tenacity import retry
+from tenacity.retry import retry_if_exception, retry_if_exception_type
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_random
+
+from spylib.constants import (
+    MAX_COST_EXCEEDED_ERROR_CODE,
+    OPERATION_NAME_REQUIRED_ERROR_MESSAGE,
+    THROTTLED_ERROR_CODE,
+    WRONG_OPERATION_NAME_ERROR_MESSAGE,
+)
+
+from spylib.exceptions import (
+    ShopifyCallInvalidError,
+    ShopifyError,
+    ShopifyExceedingMaxCostError,
+    ShopifyThrottledError,
+    not_our_fault,
+)
+
+
+class AssociatedUser(BaseModel):
+    id: int
+    first_name: str
+    last_name: str
+    email: str
+    email_verified: bool
+    account_owner: bool
+    locale: str
+    collaborator: bool
+
+
+class OfflineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+
+
+class OnlineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+    expires_in: int
+    associated_user_scope: str
+    associated_user: AssociatedUser
+
+
+class Token(ABC, BaseModel):
+    """
+    Abstract class for token objects. This should never be extended, as you
+    should either be using the OfflineToken or the OnlineToken.
+    """
+
+    store_name: str
+    scope: List[str]
+    access_token: Optional[str]
```

Comment:
Might be due to having some oauth stuff but I don't think this should be Optional.

Reply:
This was because the _get_token method was written as a method  instead of a class method requiring a bunch of parameters. 

https://github.com/SatelCreative/satel-spylib/blob/6f1d9f1d16aa0f8a47236ec7e02ba5dba30cc680/spylib/token.py#L150-L158

I will remove this from this PR based on your previous suggestions. Would you like me to open an issue to discuss the integration of the OAuth features? As I have some other thoughts on the process that might make it easier.